db/dcrpg/internal/stakestmts.go

```diff
@@ -60,6 +61,18 @@ const (
 	SelectUnspentTickets         = `SELECT id, tx_hash FROM tickets WHERE spend_type = 0 OR spend_type = -1
 		AND is_mainchain = true;`
 
+	SelectTicketsByPrice = `SELECT price,
+		SUM(CASE WHEN block_height >= $1 THEN 1 ELSE 0 END) as immature,
+		SUM(CASE WHEN block_height < $1 THEN 1 ELSE 0 END) as live FROM tickets
+		WHERE pool_status=0 AND is_mainchain=TRUE GROUP BY price ORDER BY price;`
+
+	SelectTicketsByPurchaseDate = `SELECT (block_time/$1)*$1 as timestamp,
```

Comment:
ok. My end the queries we so slow in transactions table. Too 3 secs at best

Reply:
Do we keep the `pool_status` index? My laptop is super slow without the index. I initially tried out the join on transactions and tickets table but that brought no significant performance increase.

Reply:
This query should be our worst case scenario since it returns the biggest payload and its generally been the slowest 

```sql
SELECT (block_time/1)*1 as timestamp, 
   SUM(price) as price, 
   SUM(CASE WHEN block_height >= 267622 THEN 1 ELSE 0 END) as immature,
   SUM(CASE WHEN block_height < 267622 THEN 1 ELSE 0 END) as live 
FROM tickets 
WHERE pool_status = 0 AND is_mainchain = TRUE  
GROUP BY timestamp ORDER BY timestamp;
```

Without the Index 
```
dcrdata=# explain analyze SELECT (block_time/1)*1 as timestamp, SUM(price) as price, SUM(CASE WHEN block_height >= 267622 THEN 1 ELSE 0 END) as immature, SUM(CASE WHEN block_height < 267622 THEN 1 ELSE 0 END) as live FROM tickets WHERE pool_status = 0 AND is_mainchain = TRUE  GROUP BY timestamp ORDER BY timestamp;
                                                                    QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=188362.39..193490.83 rows=30268 width=32) (actual time=5470.357..5494.540 rows=16635 loops=1)
   Group Key: (((block_time / 1) * 1))
   ->  Gather Merge  (cost=188362.39..192706.87 rows=32994 width=32) (actual time=5470.354..5486.915 rows=19750 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial GroupAggregate  (cost=187362.36..187898.52 rows=16497 width=32) (actual time=5444.093..5451.107 rows=6583 loops=3)
               Group Key: (((block_time / 1) * 1))
               ->  Sort  (cost=187362.36..187403.61 rows=16497 width=20) (actual time=5444.078..5445.483 rows=14202 loops=3)
                     Sort Key: (((block_time / 1) * 1))
                     Sort Method: quicksort  Memory: 2749kB
                     ->  Parallel Seq Scan on tickets  (cost=0.00..186206.76 rows=16497 width=20) (actual time=0.146..5435.731 rows=14202 loops=3)
                           Filter: (is_mainchain AND (pool_status = 0))
                           Rows Removed by Filter: 446455
 Planning time: 0.574 ms
 Execution time: 5495.744 ms
(15 rows)
```

With the index
```
dcrdata=# explain analyze SELECT (block_time/1)*1 as timestamp, SUM(price) as price, SUM(CASE WHEN block_height >= 267622 THEN 1 ELSE 0 END) as immature, SUM(CASE WHEN block_height < 267622 THEN 1 ELSE 0 END) as live FROM tickets WHERE pool_status = 0 AND is_mainchain = TRUE  GROUP BY timestamp ORDER BY timestamp;
                                                                     QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=32566.09..32642.24 rows=30462 width=32) (actual time=112.169..114.041 rows=16635 loops=1)
   Sort Key: (((block_time / 1) * 1))
   Sort Method: quicksort  Memory: 2068kB
   ->  HashAggregate  (cost=29840.54..30297.47 rows=30462 width=32) (actual time=103.263..107.209 rows=16635 loops=1)
         Group Key: ((block_time / 1) * 1)
         ->  Index Scan using pool_status_index on tickets  (cost=0.43..29241.46 rows=39939 width=20) (actual time=0.051..86.166 rows=42607 loops=1)
               Index Cond: (pool_status = 0)
               Filter: is_mainchain
 Planning time: 0.356 ms
 Execution time: 118.356 ms
(10 rows)
```

Without the index you can see the query takes 5 sec compared to 120 ms with the index.

Reply:
Sure, the pool_status index is fine, although it's surprising the difference is so dramatic in your machine.

The main point in my comments here are that `JOIN transactions ON purchase_tx_db_id=transactions.id` will remove the need to duplicate block_time.

Reply:
ok.  I haven't tried that out.

Reply:
Its still abit slow but not that much..
```
dcrdata=# explain analyze SELECT (tickets.block_time/86400)*86400 AS TIMESTAMP,  SUM(tickets.price) AS price,  SUM(CASE   WHEN tickets.block_height >= 269601 THEN 1 ELSE 0 END) AS immature, SUM(CASE  WHEN tickets.block_height < 269601 THEN 1  ELSE 0  END) AS live FROM tickets JOIN transactions ON purchase_tx_db_id=transactions.id WHERE pool_status = 0 AND tickets.is_mainchain = TRUE GROUP BY TIMESTAMP ORDER BY TIMESTAMP;
                                                                                   QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=159173.92..163272.42 rows=30462 width=32) (actual time=216.365..223.205 rows=142 loops=1)
   Group Key: (((tickets.block_time / 86400) * 86400))
   ->  Gather Merge  (cost=159173.92..162580.55 rows=23494 width=32) (actual time=216.356..227.112 rows=284 loops=1)
         Workers Planned: 1
         Workers Launched: 1
         ->  Partial GroupAggregate  (cost=158173.91..158937.46 rows=23494 width=32) (actual time=212.026..217.625 rows=142 loops=2)
               Group Key: (((tickets.block_time / 86400) * 86400))
               ->  Sort  (cost=158173.91..158232.64 rows=23494 width=20) (actual time=212.012..213.536 rows=21304 loops=2)
                     Sort Key: (((tickets.block_time / 86400) * 86400))
                     Sort Method: quicksort  Memory: 2741kB
                     ->  Nested Loop  (cost=0.86..156468.24 rows=23494 width=20) (actual time=0.141..204.344 rows=21304 loops=2)
                           ->  Parallel Index Scan using pool_status_index on tickets  (cost=0.43..28877.31 rows=23494 width=28) (actual time=0.081..66.663 rows=21304 loops=2)
                                 Index Cond: (pool_status = 0)
                                 Filter: is_mainchain
                           ->  Index Only Scan using transactions_pkey on transactions  (cost=0.43..5.43 rows=1 width=8) (actual time=0.006..0.006 rows=1 loops=42607)
                                 Index Cond: (id = tickets.purchase_tx_db_id)
                                 Heap Fetches: 25254
 Planning time: 0.752 ms
 Execution time: 227.584 ms
(19 rows)
```

Reply:
- We still need the `pool_status_index` as without it the Join query takes 5 sec and with it takes 200ms
- Even though the join query is slower by some milliseconds, is it fast enough for our use case?

Reply:
@chappjc Whats do we de here?

Reply:
I see in the query in your comment above it's using `tickets.block_time`, but you've done the join so you don't need that right?

Reply:
with `transactions.block_time` its even worse
```
histogram_dcrdata=# explain analyze SELECT (transactions.block_time/86400)*86400 AS TIMESTAMP,  SUM(tickets.price) AS price,  SUM(CASE   WHEN tickets.block_height >= 269601 THEN 1 ELSE 0 END) AS immature, SUM(CASE  WHEN tickets.block_height < 269601 THEN 1  ELSE 0  END) AS live FROM tickets JOIN transactions ON purchase_tx_db_id=transactions.id WHERE pool_status = 0 AND tickets.is_mainchain = TRUE GROUP BY TIMESTAMP ORDER BY TIMESTAMP;
                                                                            QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=244787.11..250085.76 rows=39821 width=32) (actual time=5256.253..5263.711 rows=143 loops=1)
   Group Key: (((transactions.block_time / 86400) * 86400))
   ->  Gather Merge  (cost=244787.11..249156.61 rows=33184 width=32) (actual time=5256.235..5267.032 rows=428 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial GroupAggregate  (cost=243787.09..244326.33 rows=16592 width=32) (actual time=5250.463..5256.348 rows=143 loops=3)
               Group Key: (((transactions.block_time / 86400) * 86400))
               ->  Sort  (cost=243787.09..243828.57 rows=16592 width=20) (actual time=5250.445..5251.930 rows=14152 loops=3)
                     Sort Key: (((transactions.block_time / 86400) * 86400))
                     Sort Method: quicksort  Memory: 1534kB
                     ->  Nested Loop  (cost=0.43..242624.14 rows=16592 width=20) (actual time=0.556..5242.289 rows=14152 loops=3)
                           ->  Parallel Seq Scan on tickets  (cost=0.00..152400.19 rows=16592 width=20) (actual time=0.195..4626.668 rows=14152 loops=3)
                                 Filter: (is_mainchain AND (pool_status = 0))
                                 Rows Removed by Filter: 447000
                           ->  Index Scan using transactions_pkey on transactions  (cost=0.43..5.43 rows=1 width=16) (actual time=0.042..0.042 rows=1 loops=42457)
                                 Index Cond: (id = tickets.purchase_tx_db_id)
 Planning time: 2.890 ms
 Execution time: 5267.554 ms
(18 rows)
```